sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala

```diff
@@ -0,0 +1,350 @@
+            val (mergedProjectList, newOutputMap) =
+              mergeNamedExpressions(np.projectList, outputMap, cp.output)
+            val mergedPlan = Project(mergedProjectList, mergedChild)
+            mergedPlan -> newOutputMap
+          }
+        case (np: Aggregate, cp: Aggregate) if supportedAggregateMerge(np, cp) =>
+          tryMergePlans(np.child, cp.child).flatMap { case (mergedChild, outputMap) =>
+            val mappedNewGroupingExpression =
+              np.groupingExpressions.map(mapAttributes(_, outputMap))
+            // Order of grouping expression doesn't matter so we can compare sets
+            if (mappedNewGroupingExpression.map(_.canonicalized) ==
+              cp.groupingExpressions.map(_.canonicalized)) {
+              val (mergedAggregateExpressions, newOutputMap) =
+                mergeNamedExpressions(np.aggregateExpressions, outputMap, cp.aggregateExpressions)
+              val mergedPlan =
+                Aggregate(cp.groupingExpressions, mergedAggregateExpressions, mergedChild)
+              Some(mergedPlan -> newOutputMap)
+            } else {
+              None
+            }
+          }
+
+        case (np: Filter, cp: Filter) =>
+          tryMergePlans(np.child, cp.child).flatMap { case (mergedChild, outputMap) =>
+            val mappedNewCondition = mapAttributes(np.condition, outputMap)
+            // Comparing the canonicalized form is required to ignore different forms of the same
+            // expression.
+            if (mappedNewCondition.canonicalized == cp.condition.canonicalized) {
+              val mergedPlan = cp.withNewChildren(Seq(mergedChild))
+              Some(mergedPlan -> outputMap)
+            } else {
+              None
+            }
+          }
+
+        case (np: Join, cp: Join) if np.joinType == cp.joinType && np.hint == cp.hint =>
+          tryMergePlans(np.left, cp.left).flatMap { case (mergedLeft, leftOutputMap) =>
+            tryMergePlans(np.right, cp.right).flatMap { case (mergedRight, rightOutputMap) =>
+              val outputMap = leftOutputMap ++ rightOutputMap
+              val mappedNewCondition = np.condition.map(mapAttributes(_, outputMap))
+              // Comparing the canonicalized form is required to ignore different forms of the same
+              // expression and `AttributeReference.quailifier`s in `cp.condition`.
+              if (mappedNewCondition.map(_.canonicalized) == cp.condition.map(_.canonicalized)) {
+                val mergedPlan = cp.withNewChildren(Seq(mergedLeft, mergedRight))
+                Some(mergedPlan -> outputMap)
+              } else {
+                None
+              }
+            }
+          }
+
+        // Otherwise merging is not possible.
+        case _ => None
+      })
+  }
+
+  private def createProject(attributes: Seq[Attribute], plan: LogicalPlan): Project = {
+    Project(
+      Seq(Alias(
+        CreateNamedStruct(attributes.flatMap(a => Seq(Literal(a.name), a))),
```

Comment:
Maybe I was not clear enough in https://github.com/apache/spark/pull/32298/files#r852171895

My idea is to not use `CreateNamedStruct` to merge subqueries, but simply let the `CTERelationDef` return multiple columns. Then we can select the corresponding column from `CTERelationRef` in each subquery, assuming the CTE reuse always works.

Reply:
Thanks. I see your point now.
Actually that idea came up already. With the merged struct I was able to reuse the whole subquery. But, without it I think `Repartition` inserts an extra exchange (that seems to be a bit an overkill in this case) and reuse only that.

As https://github.com/apache/spark/pull/34929 just got merged, I will try to rebase my PR on the top of `master` and come back to you today.